Many smart pipe fixes

esainane opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Most bridging was not happening when it needed to, and some bridging was happening against pipes on completely different layers that construction logic shouldn't have been considering.
The direction was also not being set properly.

This fixes #60978 and obsoletes #60804.

~~Note that the smart logic could stand to become smarter too, and I suspect the bridge-specific check isn't needed with the other checks already in place, though it shouldn't hurt. Turning a pipe into a bridge might now happen in cases where it doesn't need to, though running through the basic examples and a few pet projects didn't turn up anything particularly egregious.~~

Smart pipe negotiation during construction has now been completely reworked, even functioning when you're placing down a non-smart pipe that conflicts with a smart pipe.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Almost none of the setups demonstrated in the first merged PR's [BASIC PIPING GIFs](https://imgur.com/a/BDueHKa) work anymore. This fixes it so that they all work again.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
fix: Smart pipes now calculate their direction properly, and construction code checks direction logic properly. This fixes many problems with atmospherics construction.
fix: Smart pipe construction now no longer tries to work around irrelevant pipes which are on completely different layers and do not interact with the new pipe in any way.
fix: Smart pipes with less than two allowed directions can no longer be built. The pipe would be invisible, but still interact in weird ways.
fix: One tile can no longer have more than one atmospheric component ready to connect in the same direction and layer. This prevents the scenario where a temperature pump is placed in a manner which looks like it's going to connect to a horizontal bridge pipe, but actually turns the vertical plasma input pipe behind it into a manifold, ruptures the connected plasma canister, and sets Toxins on fire.
qol: Smart pipes are now much, much smarter. During construction, if a smart pipe would block placement of a component, it will see if permitted connectivity can be reconfigured to make room for other components. This includes negotiation with other smart pipes, preferring to create pipe crossings where available, or create an opposing corner pipe if not. The RPD can now also reconfigure allowed connectivity for existing smart pipes manually. Smart pipes can never be reconfigured to have less than two allowed directions, and active connections cannot be reconfigured. Automatic reconfiguration will never cause a smart pipe to permit connectivity that was not previously permitted.
fix: Multi-layer pipe components can trigger smart pipe negotiation where relevant.
refactor: Rearranged some atmospherics pipe code. Fewer redundant loops, add and use documented and centralized helpers for bit manipulation where sensible. Make bitfield assumptions explicit.
code: made a small part of the documentation in the area less wrong dear god my eyes the goggles do nothing
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

esainane committed:
Smart pipes now actually go over smart pipes

This was ostensibly the main feature of smart pipes, so should be
fairly important

Parentheses are technically not required around the |, but I think
it's important to emphasise the importance of parentheses around
bitwise operators as this was apparently broken since 3e8407c4716b

Also purge var/connections and var/connection_num, since it's
literally only used for a `machine.connection_num < 3` check
which is completely redundant, and we already build a bitfield
tracking active directions.

esainane committed:
Skip pipes on different layers earlier

This prevents a pipe being made into a bridge perpindicular to a pipe
on a completely different layer, then failing to turn into a bridge
perpindicular to a relevant pipe.

esainane committed:
Smart pipes bridge over straight pipe types

This improves bridging to now go over most devices where possible,
letting you, for example, bridge over a perpindicular layer adapter.

esainane committed:
Pipe construction: Better documentation

Cleans up smart pipe documentation to match what's actually happening.

No functional changes.

Ghilker commented:
Thanks for the work on this, I was going to check on these issues later on but I'm really glad you got it first so that I'm not the only one working on this

Ghilker commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -167,23 +167,29 @@ Buildable meters
+			if (!istype(machine, /obj/machinery/atmospherics/pipe/smart) || lowertext(machine.pipe_color) != lowertext(pipe_color))
+				var/other_direction = machine.dir
+				if(((other_direction & (EAST|WEST)) || (other_direction & (SOUTH|NORTH))) && !ISDIAGONALDIR(other_direction))
+					pipe_type = /obj/machinery/atmospherics/pipe/bridge_pipe
```
Since you can now set the direction of the pipes I think you could ditch this behaviour but uh test it before doing it

Ghilker replied:
Tho this may not work with mapped pipes, at least because my pr to set the dir of mapped pipes didn't pass

esainane replied:
Line 60 in the first commit ( https://github.com/tgstation/tgstation/pull/60981/commits/631cafe48e44dce15c82dd81326714504c203bd8#diff-8aade545476ce46e983e1a3fa0bfd6e4d2cfcee4490b8652ed644ee5044b41feR60 ) should fix dir problems in a similar way to #60804 but more concisely. I've tested and confirmed that this works with mapped pipes now

esainane replied:
I tried to stick to fairly conservative fixes for this PR, since the current situation is a pain in the ass.

I'll look into it anyway, though. If this PR is still open once I'm confident that behavior removal doesn't break anything (which might take a while there are a shitload of edge cases dear god) I'll amend it with some more cleanup. Otherwise, I'll follow up in another PR.

esainane replied:
Hmm, by being able to set the direction of the pipes now, did you mean that bridges could be replaced by a "regular" pipe going straight across? I think you might be able to smartly build a corner when there's only an opposing corner on the same layer, in a similar fashion. I'll see how that feels and will probably keep that for another PR though

Ghilker replied:
IIRC you can in theory have the pipes cross each other if their direction is different, like you can have a + cross with one pipe N|S and the other E|W, you can also have two corners on the same layer on the same tile too.
You could try to remove the auto bridge pipe thing, but you'll have to test it a bit

esainane replied:
Your command is my wish

LemonInTheDark commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -167,23 +167,29 @@ Buildable meters
-
-		if(flags & PIPING_BRIDGE && !(machine.pipe_flags & PIPING_BRIDGE) && check_ninety_degree_dir(machine)) //continue if we are placing a bridge pipe over a normal pipe only (prevent duplicates)
+		// skip checks if we don't overlap layers, either by being on the same layer or by something being on all layers
+		if((machine.piping_layer != piping_layer) && !((machine.pipe_flags | flags) & PIPING_ALL_LAYER))
```
you don't need to () wrap the first check, just adds noise

esainane replied:
Done

LemonInTheDark commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -167,23 +167,29 @@ Buildable meters
 
-		if((machine.piping_layer != piping_layer) && !((machine.pipe_flags | flags) & PIPING_ALL_LAYER)) //don't continue if either pipe goes across all layers
+		// If we're a smart pipe, and we're looking to share space with a potentially perpendicular kind of pipe, see if we can smartly go over it as a bridge
+		if(pipe_count == 1 && machine.connection_num < 3 && ispath(pipe_type, /obj/machinery/atmospherics/pipe/smart))
```
This connection_num check does nothing. I'm going to merge https://github.com/tgstation/tgstation/pull/60804 because it makes the variable actually change ever, which I assume causes issues here.

This is partially because I feel bad just closing the guy's pr cause you came along and opened a separate fix, and partially because I think it's valuable

esainane replied:
The `connection_num` check is redundant, as anything with more than two connections is going to be caught by the immediate `!ISDIAGONAL` check.

`git grep connection_num` indicates that this variable is used literally nowhere else, so the lack of variable maintenance doesn't cause issues here or anywhere else, and the variable should be safe be purge. `dir` can be updated at a point where `bitfield` is identical to `connections` and `bits` is identical to `connection_num`, so there's not much point in building two copies of everything.

#60804 on its own doesn't fix much. See the second screenshot in #60978 for how pipes are still broken with #60804 merged.

Rebased, updated, and verified that (at least) everything in [BASIC PIPING GIFs](https://imgur.com/a/BDueHKa) now works again.

esainane committed:
Pop superfluous parentheses

esainane committed:
Prevent creation of stub pipes

Setting the RPD to create a smart pipe that can only connect in one
direction would cause it to create invisible pipes that would still
block the placement of other pipes.

esainane committed:
The RPD can now reprogram smart pipes

The RPD can only affect smart pipes in directions that they are not
currently connected to.

This makes it easier to adjust designs after the fact, including
prevening round-start pipes from eagerly linking to a grey layer
adapter.

esainane committed:
Even smarter pipes

No more turning into bridge pipes. We now only try to be smart if
placement would fail, and we do our best to find a solution
whenever there is at least one smart pipe involved, regardless
of whether we're currently placing a smart pipe or not.

We never reconfigure any directions that a smart pipe is currently
connected, and we never reconfigure a smart pipe to have one or less
usable directions.

esainane committed:
Smartly go across perpindicular layer adapters

Also works when we're placing a layer adapter perpindicular to a
promiscuous smart pipe

Ghilker commented:
mate the latest changes are godly, keep it up!

esainane commented:
This feels pretty slick. I've derped around with half-remembered designs, and not only have I never needed to touch the smart pipe restriction buttons despite a lot of exotic overlaps, I think I've been putting things together faster than I was when I'd optimized setup under the old system

![smarter-pipes](https://user-images.githubusercontent.com/14355175/130638627-408bba4d-99de-4aba-9840-a20bd8ffbd5c.png)

Not heaving to leave open manifolds everywhere when you're designing things makes for much less clutter too. Even ugly ad-hoc turnarounds are pretty easily to follow like this

Ghilker commented:
Wow, i tested it on local and it feels so good to use now! One last "fix" you could do is to change the default color from grey to something different, like green or cyan

esainane commented:
Not sure what to think about that, at least in Toxins I've wanted everything to connect to everything. Maybe roundstart pipes or maintenance pipes should start with tighter permitted directions, and let people to reprogram a few pipes if they want to add a new area or something? Gating "connects waste to distro somewhere in maint" behind access to an RPD isn't that much harder than gating it behind access to a wrench. That's probably best saved for another PR though

LemonInTheDark commented:
> 
> 
> Wow, i tested it on local and it feels so good to use now! One last "fix" you could do is to change the default color from grey to something different, like green or cyan, that way there should be less "Oh i used grey, well now distro is full of plasma"

We should do this separately when the freeze ends. rename the option to like, global or all or something and change the default

LemonInTheDark commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -159,33 +159,26 @@ Buildable meters
-
-		if(flags & PIPING_BRIDGE && !(machine.pipe_flags & PIPING_BRIDGE) && check_ninety_degree_dir(machine)) //continue if we are placing a bridge pipe over a normal pipe only (prevent duplicates)
+		// skip checks if we don't overlap layers, either by being on the same layer or by something being on all layers
+		if(machine.piping_layer != piping_layer && !((machine.pipe_flags | flags) & PIPING_ALL_LAYER))
```
I don't like that you need to dupe this check. Could you make pipe_count into like a list of acceptable machines and then iterate over that, using its length where appropriate?

esainane replied:
Done

LemonInTheDark commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -159,33 +159,26 @@ Buildable meters
 			continue
 		pipe_count += 1
 	for(var/obj/machinery/atmospherics/machine in loc)
 		if((machine.pipe_flags & flags & PIPING_ONE_PER_TURF)) //Only one dense/requires density object per tile, eg connectors/cryo/heater/coolers.
 			to_chat(user, span_warning("Something is hogging the tile!"))
 			return TRUE
```
Could this be done in the other for loop?

esainane replied:
Done

LemonInTheDark commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -201,6 +194,83 @@ Buildable meters
+			// Two smart pipes. This is going to get complicated.
+			// Check to see whether the already placed pipe is bent or not.
+			if (ISDIAGONALDIR(other_smart_pipe.dir))
+				// The other pipe is bent, with at least two current connections. See if we can bounce off it as a bent pipe in the other direction.
```
No complaints, I just really like that you've implemented this bit of logic it's been bugging me for a bit

LemonInTheDark reviewed (changes requested):
Gonna need to read a little closer, some of this bitfield manip seems a bit out there, but I really like what I've seen in terms of behavior thank you

esainane committed:
Pipes: Factor out loops and some bitfiddling

Create and use helpers with documented purposes over inline bitfiddling,
when it makes sense.

Many loops and switch statements were recreating information that was
already there.

Some relationships between pipe bitfield states were already assumed.
This centralises the functionality that relies upon these assumptions,
places them where the bits are defined, and documents them.

Rewrite some bitwise operations to be more idiomatic.

esainane committed:
Smart pipes: Debugging output

I normally clean history before pushing changes to any project,
but I feel like this should be saved.

esainane committed:
Revert "Smart pipes: Debugging output"

This reverts commit bb3aa76cf6d08e4d0951113a26fc9d48b6bc1735.

esainane committed:
Add trailing comma

The lack of this was making the linter sad

stylemistake reviewed (approved):
Resolve conflicts, and you're good to go.

esainane commented:
Rebased
